docs/api.rst

```diff
@@ -3,8 +3,72 @@ Python API
 
 Pipeline
 --------
+In DALI, any data processing task has a central object called pipeline. Pipeline object is an
+instance of :class:`nvidia.dali.pipeline.Pipeline` or a derived class. Pipeline encapsulates the
+data processing graph and the execution engine.
+
+There are two ways to define a DALI pipelines:
+
+#. by inheriting from Pipeline class and overriding :meth:`nvidia.dali.pipeline.Pipeline.define_graph`
+#. by instantiating `Pipeline` directly, building the graph and setting the pipeline
+   outputs with :meth:`nvidia.dali.pipeline.Pipeline.set_outputs`
+
+Data processing graphs
+""""""""""""""""""""""
+
+DALI pipeline is represented as a graph of operations. There are two kinds of nodes in the graph:
+
+ * Operators
+ * Data nodes (see :class:`nvidia.dali.pipeline.DataNode`)
+
+The nodes are created on each call to an operator object (or function) - each call typically
+creates one operator node and as many data nodes as there are outputs from the operator.
+Passing `DataNode` objects as inputs or arguments to an operator establishes connections in
+the graph.
+Example::
+
+    class MyPipeline(Pipeline):
+        def define_graph(self):
+            img_reader  = ops.FileReader(file_root = "image_dir", seed = 1)
+            mask_reader = ops.FileReader(file_root = "mask_dir", seed = 1)
+            img_files, labels = img_reader()  # creates an instance of `FileReader`
+            mask_files, _ = mask_reader()     # creates another instance of `FileReader`
+            decode = ops.ImageDecoder()
+            images = decode(img_files)  # creates an instance of `ImageDecoder`
+            masks  = decode(mask_files)   # creates another instance of `ImageDecoder`
+            return [images, masks, labels]
+
+    pipe = MyPipeline(batch_size = 4, num_threads = 2, device_id = 0)
+    pipe.build()
+
+The resulting graph is:
+
+.. image:: images/two_readers.svg
+
+Current pipeline
+""""""""""""""""
+
+Subgraphs that do not contribute to the pipeline output are automatically pruned.
+If an operator has side effects (e.g. PythonFunction operator family), it cannot be invoked
+without setting the current pipeline. Current pipeline is set implicitly when the graph is
+defined inside derived pipeline's `define_graph` method. Otherwise, it can be set using context
+manager (`with` statement)::
+
+    pipe = dali.pipeline.Pipeline(batch_size = N, num_threads = 3, device_id = 0)
```

Comment:
You are using new functional API without introducing it earlier.

Reply:
I'm not. Functional API is the thing to be found in "dali.fn" and I'm not advertising it anywhere. As for the alternative way of creating pipelines - I think we can show that there is such an option, it doesn't harm us in any way.
BTW - even with current release, you can do:
```
def my_graph():
    reader = ops.FileReader(...)
    decoder = ops.ImageDecoder()
    return decoder(reader()[0])

pipe.define_graph = my_graph
```
so subclassing the pipeline is nothing more than a convention.

Reply:
You can, but no one knows about it as we don't have any example of this. And putting it only here only confuses users.

Reply:
I'd leave it and see which approach wins - if any. It doesn't cost us almost anything to maintain this one-line function (set_outputs).